File: src/lib/ipc/runtime.ts
import { invoke } from '@tauri-apps/api/core';

import {
  normalizeRuntimeSnapshot,
  type RuntimeSnapshot,
  type RuntimeStatus,
  type SerializedRuntimeSnapshot,
} from '../../features/runtime/models';

const runtimeCommandNames = {
  getRuntimeSnapshot: 'get_runtime_snapshot',
  initializeRuntimeState: 'initialize_runtime_state',
  pingRuntime: 'ping_runtime',
} as const;

export interface RuntimePingResponse {
  reachable: boolean;
  runtimeStatus: RuntimeStatus;
  status: string;
}

interface SerializedRuntimePingResponse {
  reachable: boolean;
  runtime_status: RuntimeStatus;
  status: string;
}

function normalizeRuntimePing(
  response: SerializedRuntimePingResponse,
): RuntimePingResponse {
  return {
    reachable: response.reachable,
    runtimeStatus: response.runtime_status,
    status: response.status,
  };
}

export async function getRuntimeSnapshot(): Promise<RuntimeSnapshot> {
  const snapshot = await invoke<SerializedRuntimeSnapshot>(
    runtimeCommandNames.getRuntimeSnapshot,
  );

  return normalizeRuntimeSnapshot(snapshot);
}

export async function initializeRuntimeState(): Promise<RuntimeSnapshot> {
  const snapshot = await invoke<SerializedRuntimeSnapshot>(
    runtimeCommandNames.initializeRuntimeState,
  );

  return normalizeRuntimeSnapshot(snapshot);
}

export async function pingRuntime(): Promise<RuntimePingResponse> {
  const response = await invoke<SerializedRuntimePingResponse>(
    runtimeCommandNames.pingRuntime,
  );

  return normalizeRuntimePing(response);
}

export { runtimeCommandNames };

File: src-tauri/src/commands/runtime.rs
use serde::Serialize;
use tauri::State;

use crate::models::runtime_snapshot::{RuntimeSnapshot, RuntimeStatus};
use crate::runtime::state::AppRuntimeState;

#[derive(Clone, Debug, PartialEq, Eq, Serialize)]
pub struct RuntimePingResponse {
    pub reachable: bool,
    pub runtime_status: RuntimeStatus,
    pub status: &'static str,
}

pub fn initialize_runtime_state_for_test(state: &AppRuntimeState) -> RuntimeSnapshot {
    state.initialize()
}

pub fn ping_runtime_for_test(state: &AppRuntimeState) -> RuntimePingResponse {
    RuntimePingResponse {
        reachable: true,
        runtime_status: state.snapshot().runtime_status,
        status: "ok",
    }
}

#[tauri::command]
pub fn get_runtime_snapshot(state: State<'_, AppRuntimeState>) -> RuntimeSnapshot {
    state.snapshot()
}

#[tauri::command]
pub fn initialize_runtime_state(state: State<'_, AppRuntimeState>) -> RuntimeSnapshot {
    initialize_runtime_state_for_test(state.inner())
}

#[tauri::command]
pub fn ping_runtime(state: State<'_, AppRuntimeState>) -> RuntimePingResponse {
    ping_runtime_for_test(state.inner())
}

#[cfg(test)]
mod tests {
    use super::ping_runtime_for_test;
    use crate::runtime::state::AppRuntimeState;

    #[test]
    fn ping_runtime_reports_host_reachability() {
        let state = AppRuntimeState::new();

        let ping = ping_runtime_for_test(&state);

        assert!(ping.reachable);
        assert_eq!(ping.status, "ok");
    }
}
